src/main/java/com/artipie/asto/Key.java

```diff
@@ -83,6 +83,21 @@ public final String string() {
         public Optional<Key> parent() {
             return this.origin.parent();
         }
+
+        @Override
+        public boolean equals(final Object another) {
+            return this.origin.equals(another);
+        }
+
+        @Override
+        public int hashCode() {
+            return this.origin.hashCode();
```

Comment:
@Sammers21 there should be additional shift for origins hash code to avoid mixing of origin and wrap in hash maps:`return 17 + origin.hashCode()`